src/eterna/mode/ThreeView.ts

```diff
@@ -0,0 +1,904 @@
+            if (this.pressed) e.stopPropagation();
+            if (!this.threeView.isOver3DCanvas) {
+                this.threeView.hideTooltip();
+            }
+        }
+    }
+
+    public handleTouchEvent(e: TouchEvent): void {
+        const pos = this.threeView.getPosition();
+        if (e.touches.length > 0) {
+            const touchObjArray = [];
+            const x = e.touches[0].clientX - pos.x;
+            const y = e.touches[0].clientY - pos.y;
+            for (let i = 0; i < e.touches.length; i++) {
+                touchObjArray.push(new Touch({
+                    identifier: Date.now(),
+                    target: this.threeView.nglGate.stage.viewer.getWebGLCanvas(),
+                    clientX: e.touches[i].clientX - pos.x,
+                    clientY: e.touches[i].clientY - pos.y,
+                    screenX: e.touches[i].screenX,
+                    screenY: e.touches[i].screenY,
+                    pageX: e.touches[i].pageX - pos.x,
+                    pageY: e.touches[i].pageY - pos.y,
+                    radiusX: e.touches[i].radiusX,
+                    radiusY: e.touches[i].radiusY,
+                    rotationAngle: e.touches[i].rotationAngle,
+                    force: e.touches[i].force
+                }));
+            }
+            const init:TouchEventInit = {
+                cancelable: true,
+                bubbles: true,
+                touches: touchObjArray,
+                targetTouches: [],
+                changedTouches: touchObjArray,
+                shiftKey: e.shiftKey,
+                ctrlKey: e.ctrlKey
+            };
+            if (this.threeView.metaState === 1) {
+                init.ctrlKey = true;
+            }
+            const touchEvent = new TouchEvent(e.type, init);
+            this.threeView.isOver3DCanvas = this.threeView.PtInCanvas(x, y);
+            if (this.threeView.isOver3DCanvas || this.pressed) {
+                if (e.type === 'touchstart') this.pressed = true;
+                this.threeView.nglGate.stage.viewer.getWebGLCanvas().dispatchEvent(touchEvent);
+            }
+            if (!this.threeView.isOver3DCanvas) {
+                if (e.type === 'touchstart') this.pressed = false;
+            }
+            if (this.pressed) e.stopPropagation();
+            if (!this.threeView.isOver3DCanvas) {
+                this.threeView.hideTooltip();
+            }
+        }
+    }
+
+    protected _boundHandleTouchEvent: (e: TouchEvent) => void;
+    protected _boundHandleMouseEvent: (e: MouseEvent | PointerEvent) => void;
+
+    protected pressed: boolean = false;
+    protected threeView:ThreeView;
+}
+
+const MaxState = 2;
+const MinState = 1;
+const NormalState = 0;
+const minH = 200;
+
+interface ViewStatus {
+    x: number;
+    y: number;
+    w: number;
+    h: number;
+}
+export default class ThreeView extends ContainerObject {
+    public nglGate: Mol3DGate;
+    public pixiContainer: HTMLDivElement;
```

Comment:
Doesn't look like this is actually used anywhere?

Reply:
I removed nglGate.